<a href="https://colab.research.google.com/github/gmihran/AyrikSistemlerIleriOlasilik/blob/master/GozdeAltinsoyOdev2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Karşılaşma Problemi (Permütasyon)

### **Problem** : Bir spor salonuna günde 3000 kişi spor yapmak için geliyor. Gelen kişilerin %40'ı kilo vermek istiyor. Sabah saatlerinde (08:00-12:00 saatleri arasında) kulübe 50 kişi geliyor. 
### Sabah saatlerinde kulübe gelen kişilerden kilo vermek isteyen bir kişiyle karşılaşma olasılığımız nedir?

**p (Toplam Kişi)** = 3000

**s (Sabah saatlerinde gelen kişi sayısı)** = 50

**f (Kilo Vermek İsteyen Kişi Sayısı)** = 3000 * 40 / 100 = 1200

*Her kişi ile karşılaşma eşit şanslı olay olursa:*

1 değerinden istenmeyen durumların gerçekleşme olasılığını çıkarırsak sonuca ulaşabiliriz.

**P (kilo vermek isteyen kişi sayısı)** = 1 - C (p-f,s) / C (p,s)

In [66]:
print("Kilo vermek isteyen kişi sayısı:",3000 * 40 / 100)

Kilo vermek isteyen kişi sayısı: 1200.0


In [0]:
def faktoriyel(sayi):
    if sayi==1:
        return 1
    else:
        return sayi * faktoriyel(sayi-1) 


def kombinasyon(n,r):
  return faktoriyel(n)/(faktoriyel(n-r)*faktoriyel(r))

In [68]:
p=3000
s=50
f=1200
k=p-f

P=1-kombinasyon(k,s)/kombinasyon(p,s)
print(P)

RecursionError: ignored

**Overflow (taşma) nedeniyle RecursionError hatası ile karşılaştık.** Bu işlemi daha küçük sayılarla yapmak mümkün olmasına rağmen, elimizdeki sayılarla recursive fonksiyonu gerçekleştirmemiz pek mümkün görünmüyor.

Recursive fonksiyonu yerine farklı bir yöntemle faktöriyel hesaplaması yapmalıyız.

Bunun için aşağıdaki faktöriyel hesaplamasını yapan fonksiyonu yazarak işlemi tekrar deniyoruz.

In [0]:
def faktoriyel(sayi):
  sonuc=1
  for i in range(1,sayi+1):
    sonuc=sonuc*i
  return sonuc
  
def kombinasyon(n,r):
  return faktoriyel(n)/(faktoriyel(n-r)*faktoriyel(r))

In [70]:
p=3000
s=50
f=1200
k=p-f

P=1-kombinasyon(k,s)/kombinasyon(p,s)
print(P)

0.9999999999938683


### Problemimizin çözümüne faktöriyel ve kombinasyon fonksiyonları ile ulaştık. Ama bu işlemlerdeki değerlere sadeleştirme işlemi uygulayarak sonuca daha hızlı bir şekilde ulaşabiliriz.
Sadeleştirme işlemi yapılarak daha kısa sürede faktöriyel hesaplanabilir.

**Bu sadeleştirmeyi aşağıdaki adımlarla yapabiliriz;**

K(n,r) kombinasyon işlemindeki n ile n-r değerleri karşılaştırırız. Bu değerlerden büyük olan değeri bölen değerle sadeleştirdikten sonra çarpım sonucunu hesaplatacağımız değerleri faktöriyel fonksiyonuna gönderiririz. Sonraki adımda ise bir önceki adımda bulmuş olduğumuz değer ile küçük olan sayının faktöriyelini birbirine böleriz.

**Ör: K(8,2) değerini ele alalım.** 

**Adım-1:** 2 ile 8-2 sayılarını karşılaştırdığımızda 6 değeri 2 değerinde daha büyük olduğu için sadeleştirme işleminde 6! sayısını kullanırız.

**Adım-2:** 8! ile 6! sayılarını sadeleştirdiğimizde elimizde 7*8 değeri kalmaktadır.Faktöriyele bu değerlerin başlangıç ve bitiş değerini gönderdiğimizde bize bu sayıların çarpımını geri döndürür.

**Adım-3:** 2! değerini hesaplatıp 2.adımda bulmuş olduğumuz 56 sonucunu 2 değerine böleriz. Böylelikle 28 değerine ulaşmış oluruz.

In [0]:
def fakt(n,r):
  sonuc=1
  for i in range(r+1,n+1):
  #r+1 den n değerine kadar faktöriyelini hesaplarız
    sonuc=sonuc*i
  return sonuc

def komb(n,r):
  nr=n-r
  if nr>r:
    K=fakt(n,nr)
    K=K/fakt(r,1)
  else:
    K=fakt(n,r)
    K=K/fakt(nr,1)
  return K

***Sağlama yapacak olursak;***

In [72]:
print(komb(8,2))

28.0


In [73]:
p=3000
s=50
f=1200
k=p-f

P=1-komb(k,s)/komb(p,s)
print(P)


0.9999999999938683


**Sonuç:** Sabah saatlerinde spor kulübüne gelen kişilerden kilo vermek isteyen bir kişiyle karşılaşma olasılığımız **0.9999999999938683** olarak bulunur.